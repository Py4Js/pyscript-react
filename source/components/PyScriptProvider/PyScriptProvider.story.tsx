import { ComponentMeta, ComponentStory } from "@storybook/react";
import { FunctionComponent } from "react";
import PyScriptProvider from "~components/PyScriptProvider/PyScriptProvider";
import { PyScriptProviderProperties } from "~components/PyScriptProvider/PyScriptProvider.types";

export default {
  title: "PyScriptProvider",
  component: PyScriptProvider,
} as ComponentMeta<typeof PyScriptProvider>;

const Template: ComponentStory<
  FunctionComponent<PyScriptProviderProperties>
> = ({ ...rest }: PyScriptProviderProperties): JSX.Element => {
  return (
    <PyScriptProvider {...rest}>
      {/* eslint-disable-next-line react/no-unescaped-entities*/}
      <py-script>print("Hello World!")</py-script>
    </PyScriptProvider>
  );
};

export const PyScriptProviderExample: ComponentStory<typeof PyScriptProvider> =
  Template.bind({});

PyScriptProviderExample.args = {
  cssSrc: "https://pyscript.net/alpha/pyscript.css",
  jsSrc: "https://pyscript.net/alpha/pyscript.js",
};
