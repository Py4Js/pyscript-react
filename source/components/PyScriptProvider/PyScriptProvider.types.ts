import {
  DetailedHTMLProps,
  LinkHTMLAttributes,
  PropsWithChildren,
  ScriptHTMLAttributes,
} from "react";
import { HelmetProps, ProviderProps } from "react-helmet-async";

export type PyScriptProviderPropertiesWithCSSSourceFullname = {
  cssSource?: string;
  cssSrc?: never;
};

export type PyScriptProviderPropertiesWithoutCSSSourceFullname = {
  cssSource?: never;
  cssSrc?: string;
};

export type PyScriptProviderPropertiesWithJSourceFullname = {
  jsSrc?: never;
  jsSource?: string;
};

export type PyScriptProviderPropertiesWithoutJSourceFullname = {
  jsSrc?: string;
  jsSource?: never;
};

export type PyScriptProviderPropertiesWithoutCSSPropertiesFullname = {
  cssProps?: Omit<
    DetailedHTMLProps<LinkHTMLAttributes<HTMLLinkElement>, HTMLLinkElement>,
    "href"
  >;
  cssProperties?: never;
};

export type PyScriptProviderPropertiesWithCSSPropertiesFullname = {
  cssProps?: never;
  cssProperties?: Omit<
    DetailedHTMLProps<LinkHTMLAttributes<HTMLLinkElement>, HTMLLinkElement>,
    "href"
  >;
};

export type PyScriptProviderPropertiesWithJSPropertiesFullname = {
  jsProps?: never;
  jsProperties?: Omit<
    DetailedHTMLProps<
      ScriptHTMLAttributes<HTMLScriptElement>,
      HTMLScriptElement
    >,
    "src"
  >;
};

export type PyScriptProviderPropertiesWithoutJSPropertiesFullname = {
  jsProps?: Omit<
    DetailedHTMLProps<
      ScriptHTMLAttributes<HTMLScriptElement>,
      HTMLScriptElement
    >,
    "src"
  >;
  jsProperties?: never;
};

export type PyScriptProviderPropertiesWithoutHelmetPropertiesFullname = {
  helmetProps?: HelmetProps;
  helmetProperties?: never;
};

export type PyScriptProviderPropertiesWithHelmetPropertiesFullname = {
  helmetProps?: never;
  helmetProperties?: HelmetProps;
};

export type PyScriptProviderPropertiesWithHelmetProviderPropertiesFullname = {
  helmetProviderProps?: never;
  helmetProviderProperties?: ProviderProps;
};

export type PyScriptProviderPropertiesWithoutHelmetProviderPropertiesFullname =
  {
    helmetProviderProps?: ProviderProps;
    helmetProviderProperties?: never;
  };

export type PyScriptProviderProperties = PropsWithChildren<
  (
    | PyScriptProviderPropertiesWithCSSSourceFullname
    | PyScriptProviderPropertiesWithoutCSSSourceFullname
  ) &
    (
      | PyScriptProviderPropertiesWithJSourceFullname
      | PyScriptProviderPropertiesWithoutJSourceFullname
    ) &
    (
      | PyScriptProviderPropertiesWithCSSPropertiesFullname
      | PyScriptProviderPropertiesWithoutCSSPropertiesFullname
    ) &
    (
      | PyScriptProviderPropertiesWithJSPropertiesFullname
      | PyScriptProviderPropertiesWithoutJSPropertiesFullname
    ) &
    (
      | PyScriptProviderPropertiesWithHelmetPropertiesFullname
      | PyScriptProviderPropertiesWithoutHelmetPropertiesFullname
    ) &
    (
      | PyScriptProviderPropertiesWithHelmetProviderPropertiesFullname
      | PyScriptProviderPropertiesWithoutHelmetProviderPropertiesFullname
    )
>;
