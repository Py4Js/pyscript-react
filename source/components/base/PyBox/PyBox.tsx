import propTypes from "prop-types";
import {
  DetailedHTMLProps,
  HTMLAttributes,
  PropsWithChildren,
  WeakValidationMap,
} from "react";
import ReactElementProps from "~types/ReactElementProps/ReactElementProps";

export type PyBoxPropertiesBase = PropsWithChildren<
  ReactElementProps<DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>>
>;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type PyBoxProperties<T> = T extends infer T
  ? T & PyBoxPropertiesBase
  : PyBoxPropertiesBase;

export type PyBoxTag = {
  <T extends object>(properties: PyBoxProperties<T>): JSX.Element;
  propTypes: WeakValidationMap<PyBoxPropertiesBase>;
};

const PyBox: PyBoxTag = <T extends object>({
  children,
  ...rest
}: PyBoxProperties<T>): JSX.Element => {
  return <py-box {...rest}>{children}</py-box>;
};

PyBox.propTypes = {
  children: propTypes.oneOfType([
    propTypes.arrayOf(propTypes.node),
    propTypes.node,
  ]).isRequired,
};

export default PyBox;
