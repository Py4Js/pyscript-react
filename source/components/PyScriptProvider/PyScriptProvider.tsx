import propTypes from "prop-types";
import { FC, memo, WeakValidationMap } from "react";
import { Helmet, HelmetProvider } from "react-helmet-async";
import { PyScriptProviderProperties } from "~components/PyScriptProvider/PyScriptProvider.types";

const PyScriptProvider: FC<PyScriptProviderProperties> = ({
  cssSrc = "https://pyscript.net/alpha/pyscript.css",
  cssSource,
  jsSrc = "https://pyscript.net/alpha/pyscript.js",
  jsSource,
  children,
  cssProps,
  cssProperties,
  jsProperties,
  jsProps,
  helmetProps,
  helmetProperties,
  helmetProviderProps,
  helmetProviderProperties,
}: PyScriptProviderProperties): JSX.Element => {
  return (
    <HelmetProvider {...(helmetProviderProperties || helmetProviderProps)}>
      <Helmet {...(helmetProperties || helmetProps)}>
        <link
          rel="stylesheet"
          {...(cssProperties || cssProps)}
          href={cssSource || cssSrc}
        />
        <script defer {...(jsProperties || jsProps)} src={jsSource || jsSrc} />
      </Helmet>
      {children}
    </HelmetProvider>
  );
};

PyScriptProvider.propTypes = {
  cssSrc: propTypes.string,
  jsSrc: propTypes.string,
  cssSource: propTypes.string,
  jsSource: propTypes.string,
  children: propTypes.oneOfType([
    propTypes.arrayOf(propTypes.node),
    propTypes.node,
  ]).isRequired,
} as WeakValidationMap<PyScriptProviderProperties>;

export default memo(PyScriptProvider);
