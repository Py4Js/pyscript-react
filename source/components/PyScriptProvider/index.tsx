import PyScriptProvider from "~components/PyScriptProvider/PyScriptProvider";

export default PyScriptProvider;
