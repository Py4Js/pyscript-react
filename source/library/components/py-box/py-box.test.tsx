import { render } from "@testing-library/react";
import PyScriptProvider from "../py-script-provider";
import PyBox from "./py-box";

const children = [
  <div data-testid="first-element" key="first-element">
    first
  </div>,
  <div data-testid="second-element" key="second-element">
    second
  </div>,
  <div data-testid="third-element" key="third-element">
    third
  </div>,
];
describe("PyBox", () => {
  it("check if match snapshots without pyscript provider", () => {
    const { container } = render(<PyBox />);
    expect(container).toMatchSnapshot();
  });
  it("check if match snapshots with alpha pyscript provider", () => {
    const { container } = render(
      <PyScriptProvider
        jsSource="https://pyscript.net/alpha/pyscript.js"
        cssSource="https://pyscript.net/alpha/pyscript.css"
      >
        <PyBox widths="2/3;1/6;1/6" children={children} />
      </PyScriptProvider>,
    );
    expect(container).toMatchSnapshot();
  });
  it("ensure that PyBox doesn't work with latest pyscript provider ", () => {
    const { container } = render(
      <PyScriptProvider>
        <PyBox children={children} />
      </PyScriptProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
