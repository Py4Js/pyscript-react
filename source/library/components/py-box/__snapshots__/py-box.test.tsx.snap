// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PyBox check if match snapshots with alpha pyscript provider 1`] = `
<div>
  <py-box
    widths="2/3;1/6;1/6"
  >
    <div
      data-testid="first-element"
    >
      first
    </div>
    <div
      data-testid="second-element"
    >
      second
    </div>
    <div
      data-testid="third-element"
    >
      third
    </div>
  </py-box>
</div>
`;

exports[`PyBox check if match snapshots with set width 1`] = `
<div>
  <py-box
    widths="2/3;1/6;1/6"
  >
    <div
      data-testid="first-element"
    >
      first
    </div>
    <div
      data-testid="second-element"
    >
      second
    </div>
    <div
      data-testid="third-element"
    >
      third
    </div>
  </py-box>
</div>
`;

exports[`PyBox check if match snapshots without pyscript provider 1`] = `
<div>
  <py-box />
</div>
`;

exports[`PyBox ensure that PyBox doesn't work with latest pyscript provider  1`] = `
<div>
  <py-box>
    <div
      data-testid="first-element"
    >
      first
    </div>
    <div
      data-testid="second-element"
    >
      second
    </div>
    <div
      data-testid="third-element"
    >
      third
    </div>
  </py-box>
</div>
`;
